### ISSUE TITLE
Disk I/O throttling?

### ISSUE THREAD

**Issue by benjamin-thomas** [restic/restic#1145]
Thanks for your great program, I find it a pleasure to use!

I've had a problem with a particular backup though...

## Output of `restic version`

    restic 0.7.1
    compiled with go1.8.3 on linux/amd64

The OS is : Ubuntu 14.04 LTS
The filesystem is: ext4

## How did you start restic exactly? (Include the complete command line)

     envdir ~/.envdir/restic/ restic backup ~/my_dir

Note: envdir provided by [daemontools](http://thedjbway.b0llix.net/daemontools/envdir.html)

## What backend/server/service did you use?

rest-server over HTTPS

## Expected behavior

Initial backup of a big directory (230GB) should not render the system unresponsive. 

## Actual behavior

Stats of the running output, for reference:

    scan [/home/user/my_dir]
    scanned 7408 directories, 5106240 files in 5:45
    [1:42:37] 52.03%  20.227 MiB/s  121.623 GiB / 233.752 GiB  448727 / 5113648 items  0 errors  ETA 1:34:36

The system became unresponsive because of a very high %CPU "wait average" (=70), available via top.

Also I confirmed that only the restic process was choking the disk via `iotop`.

## Steps to reproduce the behavior

Launch the initial backup subcommand on a folder containing millions of little files (I suppose this is the problem)

## Do you have any idea what may have caused this?

The folder contains lots of small text files (eml), segmented into subdirectories based upon year/month/day to facilitate normal disk lookups (via /bin/ls, etc).

I guess this has a negative impact on restic doing lots of disk lookups ?

For the time being, I found a workaround by wrapping the restic command as such:

    envdir ~/.envdir/restic/ nice -n19 ionice -c3 restic backup ~/my_dir

For `ionice` to work however, I also had to switch to a different IO scheduler as such:

```console
root@server:~# cat /sys/block/sda/queue/scheduler
noop [deadline] cfq
root@server:~# echo cfq > /sys/block/sda/queue/scheduler
root@server:~# cat /sys/block/sda/queue/scheduler
noop deadline [cfq]  
```
It'd be nice if there was a way to throttle I/O operations.

Even, from a usability perspective.

Because, even if this workaround works, I still get a huge load average that I have to ignore for the moment.

```console
root@server:~# cat /proc/loadavg
11.38 11.10 10.84 2/355 30862
```
My server *is* responsive at the moment, but if I get a CPU or memory spike, I wouldn't be able to see this because of the mudded loadavg.

Any thoughts?

**Comment by fd0** [comment]
Hey, we're aware of the issue: The archiver code does too many file operations in parallel. Once I get around to reworking the archiver (it's already planned) that'll get better. Stay tuned.

**Comment by benjamin-thomas** [comment]
Alright, thanks for the feedback!

**Comment by benjamin-thomas** [comment]
Adding more info for reference/documentation purpuses: no need to reply!

---

In reference to issue #329, I wanted to document my experience as I happened to test this tool by pure coincidence.

In my usage scenario, `attic` turned out not to hog the system on the whole, but also is much much slower.

With my current setup, I was able to push 300GB in 10 hours with restic on the first run.
With `attic`, I stopped my test after 24 hours: where I had only 100GB pushed.

So restic was 7 times faster, I thought you'd like to know.

Also, compared to a simple rsync, there was about a 10-20% overhead when using restic on the first run, which is not bad at all.

On second runs, there is a much bigger difference.

restic completes it's incremental backup in about 45 min (the data grows up to 200MB daily for about 10000 files): most of the time is spent on scanning the tree for changes.

With rsync, syncing completes in 5 minutes. 80-90% of the time is spent on building the list of files to transfer.

Also, prior to this problem, I had tested restic via 3 different backends (sftp, rest/https, s3 via minio), and concluded that there was no noticeable difference in speed with either one (in my use case).

Apart for soft links which the sftp backend seemed to handle better (sftp and s3 backends seemed to duplicate the data rendering the backup slower for folders containing lots of symlinks).

Keep up the good work! :)

**Comment by fd0** [comment]
Interesting observations, thanks!

> restic completes it's incremental backup in about 45 min (the data grows up to 200MB daily for about 10000 files): most of the time is spent on scanning the tree for changes.

That'll probably get better once we have a local cache for the metadata. Which backend was that?

> Apart for soft links which the sftp backend seemed to handle better (sftp and s3 backends seemed to duplicate the data rendering the backup slower for folders containing lots of symlinks).

That's odd, do you have any data on your observation? The backends are no different on that layer. Symlinks are handled, well, as symlinks, so no data is transferred apart from the metadata (symlink name, target etc.) Also, what are "soft links"? Another word for symlink?

**Comment by benjamin-thomas** [comment]
Oops sorry, yeah soft links is the same as symlinks.

But a bit tired though, I actually meant "hard links", not soft links at all.

Before `restic`, I was using [rsnapshot](http://rsnapshot.org/) which uses the concept of hard links to save on space.

Basically, you can have multiple file paths pointing to the same inode, thus saving space (as long as you're on the same file system).

It's when I was testing restic to back up my snapshot folder generated by `rsnapshot` that I noticed this difference between backends.

That "snapshot" folder contains a ton of hard links, due to the design of `rsnapshot`, but is not related to the backup I've had trouble with.

I didn't do any extensive testing, because I had to quickly change backup strategy due to a mandatory  and planned server migration. I knew my backup strategy was not appropriate anymore though so tested `restic` which I found fitted my needs much better.

I opted for the REST over HTTPS backend in the end, because I found it to be the most flexible for my use case.

**Comment by fd0** [comment]
This will be resolved by #1494, which reduces the IO to reading two files in parallel at most.

**Comment by benjamin-thomas** [comment]
Awesome, can’t wait! ;)

**Comment by fd0** [comment]
This has been resolved in #1494, I forgot to close this issue. Thanks!
